### PR TITLE
[FLINK-30449][sql-gateway] Introduce ResultInfo to improve the serialization and deserialization of FetchResultsResponseBody

### PR DESCRIPTION
**Pull request by yuzelin**
## What is the purpose of the change
Currently, the REST endpoint serialize and deserialize the `ResultSet` in the `FetchResultsResponseBody`. But many information of ResultSet will be also put in the FetchResultsResponseBody, so it's not a good way. I propose to introduce a `ResultInfo` to replace the ResultSet in the FetchResultsResponseBody.

## Brief change log
  - Add `ResultInfo` in sql gateway.
  - Add `ResultInfoJsonSerializer` and `ResultInfoJsonDeserializer` and test.
  - Remove useless ser/de code of ResultSet.

## Verifying this change

This change added tests and can be verified as follows:
  - `ResultInfoJsonSerDeTest`.
  - `SqlGatewayRestEndpointStatementITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented?  JavaDocs

### CODE REVIEW

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution.

File: flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/rest/serde/ResultInfoJsonSerDeTest.java [hunk 1]
```diff
@@ -75,80 +79,73 @@
...
-    void testResultInfoSerDeWithNullValues() throws Exception {
+    public void testResultInfoSerDeWithNullValues() throws Exception {
         List<Row> rowList = new ArrayList<>();
         List<Integer> positions = new ArrayList<>();
         for (int i = 0; i < 18; i++) {
             positions.add(new Random().nextInt(18));
         }
         for (int i = 0; i < rowNumber; i++) {
-            rowList.add(getTestRowDataWithNullValues(positions));
+            rowList.add(getTestRowDataWithNullValues(testRow, positions));
```
Comment (fsk119) [inline flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/rest/serde/ResultInfoJsonSerDeTest.java:118]: It always sets the same row.

**Comment by fsk119** [comment]
I add a commit to improve the codes. @yuzelin Please take a look.

**Comment by yuzelin** [comment]
@fsk119 thx, LGTM. I have added a bytes case to the resultInfo.txt.

**Review by fsk119 — APPROVED** [review]
LGTM

**Pull request merged**